src/components/ars.js

```diff
@@ -1,105 +1,132 @@
+    onError: (response: *) => response,
+    onFetch: (data: *) => data,
+    onChange: () => {},
+    rootAttributes: {},
+    multiselect: false,
+    resource: 'Photo',
+    mode: 'gallery',
+    url: ''
+  }
 
-  getInitialState() {
-    let { picked } = this.props
-    if (picked && !Array.isArray(picked)) {
-      picked = [picked]
-    }
+  constructor(props: Props, context: *) {
+    super(props, context)
 
-    return {
+    this.state = {
       dialogOpen: false,
-      picked: picked
+      picked: [].concat(props.picked || [])
     }
-  },
+  }
+
+  syncProps() {
+    let { makeURL, makeQuery, onError, onFetch, url } = this.props
+    return { makeURL, makeQuery, onError, onFetch, url }
+  }
 
   getPicker() {
     let { picked } = this.state
     let { columns, multiselect, mode } = this.props
 
     return (
       <Picker
-        key="dialog"
-        ref="picker"
         {...this.syncProps()}
-        onChange={this._onGalleryPicked}
-        onExit={this._onExit}
+        onChange={this._onGalleryPicked.bind(this)}
+        onExit={this._onExit.bind(this)}
         picked={picked}
         mode={mode}
         multiselect={multiselect}
         columns={columns}
       />
     )
-  },
+  }
 
   render() {
-    const { multiselect, resource, rootAttributes } = this.props
-    const { dialogOpen, picked } = this.state
-
-    let SelectionComponent = multiselect ? MultiSelection : Selection
-    let ref = multiselect ? 'multiselection' : 'selection'
-    let slug = this.props.multiselect ? picked : picked && picked[0]
-    const rootClass = cx('ars', rootAttributes.className)
+    let { multiselect, resource, rootAttributes } = this.props
+    let { dialogOpen, picked } = this.state
+
+    let rootClass = cx('ars', rootAttributes.className)
+
+    if (multiselect) {
+      return (
+        <div {...rootAttributes} className={rootClass}>
+          <MultiSelection
+            {...this.syncProps()}
+            resource={resource}
+            slugs={picked}
+            onClick={this._onOpenClick.bind(this)}
+          />
+          {dialogOpen && this.getPicker()}
+        </div>
+      )
+    }
 
     return (
       <div {...rootAttributes} className={rootClass}>
-        <SelectionComponent
-          ref={ref}
+        <Selection
           {...this.syncProps()}
           resource={resource}
-          onClick={this._onOpenClick}
-          slug={slug}
+          slug={picked && picked[0]}
+          onClick={this._onOpenClick.bind(this)}
         />
         {dialogOpen && this.getPicker()}
       </div>
     )
-  },
+  }
 
   _triggerChange() {
     let { picked } = this.state
     this.props.onChange(this.props.multiselect ? picked : picked[0])
-  },
+  }
 
   _onOpenClick() {
     this.setState({ dialogOpen: true })
-  },
+  }
 
-  _onGalleryPicked(picked) {
-    this.setState({ picked }, this._triggerChange)
-  },
+  _onGalleryPicked(picked: *) {
+    this.setState({ picked }, this._triggerChange.bind(this))
```

Comment:
just curious, what do you think of rewriting these callbacks as instance functions so you avoid `bind(this)`. Doesn't really matter, but I found a preference in that style.

```js
_triggerChange = () => {}
\\..
this.setState({ picked }), this._triggerChange)
```

Reply:
Thanks for the prod. I think this is a great idea. I'm just about done moving everything over to the updated React API. Do you mind if I do this in one swoop?